backend-rust/src/graphql_api.rs

```diff
@@ -1105,22 +1104,110 @@ impl Account {
 
     /// Number of transactions where this account is used as sender.
     async fn transaction_count<'a>(&self, ctx: &Context<'a>) -> ApiResult<i64> {
-        let rec = sqlx::query!("SELECT COUNT(*) FROM transactions WHERE sender=$1", self.index)
-            .fetch_one(get_pool(ctx)?)
-            .await?;
-        Ok(rec.count.unwrap_or(0))
+        let count =
+            sqlx::query_scalar!("SELECT COUNT(*) FROM transactions WHERE sender=$1", self.index)
+                .fetch_one(get_pool(ctx)?)
+                .await?;
+        Ok(count.unwrap_or(0))
     }
 
     async fn tokens(
         &self,
-        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: i32,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: Option<u64>,
         #[graphql(desc = "Returns the elements in the list that come after the specified cursor.")]
-        after: String,
-        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: i32,
+        after: Option<String>,
+        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: Option<u64>,
         #[graphql(desc = "Returns the elements in the list that come before the specified cursor.")]
-        before: String,
+        before: Option<String>,
     ) -> ApiResult<connection::Connection<String, AccountToken>> {
-        todo_api!()
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+        let query = ConnectionQuery::<i64>::new(
+            first,
+            after,
+            last,
+            before,
+            config.contract_connection_limit,
+        )?;
+
+        // Tokens with 0 balance are filtered out. We still display tokens with a
+        // negative balance (buggy cis2 smart contract) to help smart contract
+        // developers to debug their smart contracts.
+        // The tokens are sorted/filtered by the newest tokens transferred to an account
+        // address.
+        let mut row_stream = sqlx::query_as!(
+            AccountTokenInterim,
+            "
+            SELECT * FROM (
+                SELECT * FROM (
+                    SELECT
```

Comment:
Since we are not using row_number any more we can merge the two inner subqueries into one,